nginx.md

```diff
@@ -0,0 +1,190 @@
+
+## PHP Configuration
+
+This is a topic outside the realm of this document.  Unlike Apache (which sends every document to PHP automatically), Nginx is more streamlined.  It needs to be *told* to send a file to an external PHP processor.  It does not handle PHP itself.
+
+Configuring PHP is a topic that is largely up to the OS you are using.  The examples I'm going to provide here apply to Ubuntu 14.04, but they also require external configuration of `php-fpm` or another php processor.
+
+Your PHP configuration will look something like this:
+
+```
+location ~ [^/]\.php(/|$) {
+	fastcgi_split_path_info ^(.+?\.php)(/.*)$;
+
+	# Protection Against "cgi.fix_pathinfo = 1"
+	if (!-f $document_root$fastcgi_script_name) {
+		return 404;
+	}
+
+	fastcgi_pass unix:/var/run/php5-fpm.sock;
+	fastcgi_index index.php;
+	include fastcgi_params;
+
+	# Inform Pico we will be rewriting URL's for it.
+	fastcgi_param PICO_URL_REWRITING 1;
+}
+```
+
+Please note that this is only provided as an **example**.  You should write your own PHP location block based on your personal system configuration.
+
+This `location` rule tells Nginx to send all pages ending in `.php` to an external php processor called `php-fpm`.  Again, setting this up is outside the scope of this document.  There are many tutorials available online.  Here is one for [Ubuntu 14.04](https://www.digitalocean.com/community/tutorials/how-to-install-linux-nginx-mysql-php-lemp-stack-on-ubuntu-14-04#3-install-php-for-processing).
+
+By default, `php-fpm` comes with a very insecure setting that can allow unauthorized code execution.  We've included a small `if` statement here that will protect you from this vulnerability.  If you've changed php-fpm's `cgi.fix_pathinfo` setting to `0`, you do not need this statement.
+
+The line `fastcgi_param PICO_URL_REWRITING 1;` informs Pico that we will be rewriting url's in Nginx.  This prevents Pico from using its own url-rewriting, and gives us nicer url's.
+
+## URL Rewriting
+
+There are several ways to approach Pico's url-rewriting in Nginx.  We'll be covering two different approaches in this document, but there really is no right answer.
+
+### Pico in Document Root
+
+If your Pico installation is in the `Document Root` of your website, then configuration is relatively easy.
+
+```
+location / {
+	try_files $uri $uri/ /?$uri&$args;
+}
+```
+
+This rule tells Nginx that whenever it looks up a url within your site, it should first look for a file of that name, then a directory, and finally, if neither exist, it should rewrite the url for Pico.  When the url is rewritten for Pico, it is formatted as `/` followed by a query string (starts with `?`) containing the file path (`$uri`) and then followed by any other query arguments (`$args`).  By referencing `/`, Nginx will load `index.php` (as defined in our server config above) and pass the rendering along to your php processor.
```

Comment:
> By referencing `/`, Nginx will load `index.php`

Reading this, we can save a round trip by directly specifying `/index.php?$uri&$args`.